vnext/src/tsconfig.json

```diff
@@ -14,7 +14,8 @@
     "moduleResolution": "node",
     "noUnusedLocals": true,
     "strict": true,
-    "rootDir": "."
+    "rootDir": ".",
+    "esModuleInterop": true
```

Comment:
This could be breaking for code relying on previous esModule CommonJS interop behavior (e.g. import * is not callable with this enabled). What led to the need for the change and is it possible to change imports/exports to get the same effect? Note that TS has some special syntax for CommonJS interop as well. 

The rest of the packages in the repo have this turned off, so having consistent behavior would be nice.

Reply:
All the tests were originally authored over in packages\e2etest with this turned on.  So for example this line of code in AccessibilityTestPage.tsx won't compile without it turned on:

import React, { useState } from 'react';   => error TS1259: Module '"D:/rnw/node_modules/@types/react/index"' can only be default-imported using the 'esModuleInterop' flag

I honestly don't understand the error so I went the easy route and turned the option on.  I'm happy to change the import if only I could understand, it's quite bizarre!

---
In reply to: [457675352](https://github.com/microsoft/react-native-windows/pull/5555#discussion_r457675352) [](ancestors = 457675352)

Reply:
Reading this article over:  [https://itnext.io/great-import-schism-typescript-confusion-around-imports-explained-d512fc6769c2](https://itnext.io/great-import-schism-typescript-confusion-around-imports-explained-d512fc6769c2)

The author recommends turning this option on:  "...So I really advocate the usage of esModuleInterop option, cause it not only allows you to write less code, easier to read and spec-complaint code (and it's not just ranting, rollup for example won't allow you to use star imports in such way), but also mitigates some friction between typescript and babel communities."


---
In reply to: [457727708](https://github.com/microsoft/react-native-windows/pull/5555#discussion_r457727708) [](ancestors = 457727708,457675352)

Reply:
Yeah, Babel having it on by default and Typescript having it off by default is annoying. I'd kind of rather we have it off everywhere or on everywhere personally, and the risk of runtime issues makes me hesitant to turn it on anywhere where it uses to be off.

The root cause of the issue is that CommonJS is still used internally for React, and Node, but ES6 modules added default imports which don't really have a commonJS equivalent. The interop setting sort of changes how Typescript or Babel projects a default import. My own preference has been to just avoid default import/export with anything interfacing with CommonJS since behavior is settings dependent. 

I think the non esModuleInterop way to do this would be to have "import * as React from 'react'" on its own line instead of the default import.

Reply:
Thanks for the tip.  I turned the option off and fixed up all the imports to do it the "non-esModuleInterop" way.  Would love to read an article that actually explains all of this, I feel like I came to the party after everyone else already is in the club, leaving me feeling rather clueless.

---
In reply to: [457742505](https://github.com/microsoft/react-native-windows/pull/5555#discussion_r457742505) [](ancestors = 457742505)

Reply:
I think the historical context of being used to CommonJS before ES6 helps. Finding an article explaining CommonJS, then looking at the ES6 article on MDN should give some context on the different models, and why default imports (or setting module.exports to a single function) are wonky.

The basic idea is that CommonJS modules export a single value. This is often an object that's an aggregate of exports, which maps directly to most of ES6, but ES6 default imports don't really have a CommonJS equivalent. Without esModuleInterop, a default import will look for the ".default" property of the CommonJS export, and agregate imports are callable, since CommonJS allows exporting single functions.

With interop on, idefault imports will import the entire aggregate object IIRC, and calling a CommonJS export requires doing a default import instead of importing all.